js/moz-parser.js

```diff
@@ -0,0 +1,149 @@
+          doAddSection(lastSection);
+          lastSection.code = '';
+        }
+        for (const f of e.functions) {
+          const m = f && f.match(/^([\w-]*)\((['"]?)(.+?)\2?\)$/);
+          if (!m || !/^(url|url-prefix|domain|regexp)$/.test(m[1])) {
+            errors.push(`${e.line}:${e.col + 1} invalid function "${m ? m[1] : f || ''}"`);
+            continue;
+          }
+          const aType = CssToProperty[m[1]];
+          const aValue = aType !== 'regexps' ? m[3] : m[3].replace(/\\\\/g, '\\');
+          (section[aType] = section[aType] || []).push(aValue);
+        }
+        sectionStack.push(section);
+      });
+
+      parser.addListener('enddocument', function () {
+        const end = backtrackTo(this, parserlib.css.Tokens.RBRACE, 'start');
+        const section = sectionStack.pop();
+        const lastSection = sectionStack[sectionStack.length - 1];
+        section.code += getRange(section.start, end);
+        lastSection.start = (++end.col, end);
+        doAddSection(section);
+      });
+
+      parser.addListener('endstylesheet', () => {
+        // add nonclosed outer sections (either broken or the last global one)
+        const lastLine = lines[lines.length - 1];
+        const endOfText = {line: lines.length, col: lastLine.length + 1};
+        const lastSection = sectionStack[sectionStack.length - 1];
+        lastSection.code += getRange(lastSection.start, endOfText);
+        sectionStack.forEach(doAddSection);
+
+        if (errors.length) {
+          reject(errors);
+        } else {
+          resolve(sections);
+        }
+      });
+
+      parser.addListener('error', e => {
+        errors.push(e.line + ':' + e.col + ' ' +
+          e.message.replace(/ at line \d.+$/, ''));
+      });
+
+      parser.parse(mozStyle);
+
+      function getRange(start, end) {
+        const L1 = start.line - 1;
+        const C1 = start.col - 1;
+        const L2 = end.line - 1;
+        const C2 = end.col - 1;
+        if (L1 === L2) {
+          return lines[L1].substr(C1, C2 - C1 + 1);
+        } else {
+          const middle = lines.slice(L1 + 1, L2).join('\n');
+          return lines[L1].substr(C1) + '\n' + middle +
+            (L2 >= lines.length ? '' : ((middle ? '\n' : '') + lines[L2].substring(0, C2)));
+        }
+      }
+
+      function doAddSection(section) {
+        section.code = section.code.trim();
+        // don't add empty sections
+        if (
+          !section.code &&
+          !section.urls &&
+          !section.urlPrefixes &&
+          !section.domains &&
+          !section.regexps
+        ) {
+          return;
+        }
+        /* ignore boilerplate NS */
+        if (section.code === '@namespace url(http://www.w3.org/1999/xhtml);') {
+          return;
+        }
+        sections.push(Object.assign({}, section));
+      }
+    });
+  }
+
+  return {
+    // Parse mozilla-format userstyle into sections
+    parse(text) {
+      if (typeof parserlib === 'undefined') {
```

Comment:
Is there reall a need for this if-else? Simply always use loadScript + then
The overhead to check the cache is negligible, like 1 microsecond.

Reply:
Sometime the script is already loaded.
https://github.com/openstyles/stylus/blob/a8796b0bd2799ad3ac67e3dab663c3d4bbb2a3d0/edit/edit.js#L1643
This should be fixed when switching from loadScript to onDOMscripted.